### PR TITLE
meilisearch: 0.23.1 -> 0.28.1

### PR DESCRIPTION
**Pull request by happysalada**
###### Description of changes

testing out a build for an upstream update.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@GrahamcOfBorg eval

**Comment by happysalada** [comment]
@GrahamcOfBorg eval

**Comment by happysalada** [comment]
for some reason I'm hitting a weird compilation error on linux.
I'm going to leave this open and just try again on the next version.

**Comment by FliegendeWurst** [comment]
The newest version fails to compile because it tries to download data at build time:
```
error: failed to run custom build command for `lindera-ipadic v0.12.6`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/lindera-ipadic-25c81d6fb8bc2bba/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-changed=Cargo.toml

  --- stderr
  Error: Transport(Transport { kind: Dns, message: Some("resolve dns name 'jaist.dl.sourceforge.net:80'"), url: Some(Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("jaist.dl.sourceforge.net")), port: None, path: "/project/mecab/mecab-ipadic/2.7.0-20070801/mecab-ipadic-2.7.0-20070801.tar.gz", query: None, fragment: None }), source: Some(Custom { kind: Uncategorized, error: "failed to lookup address information: Temporary failure in name resolution" }) })
```

This is already fixed in lindera 0.13 (https://github.com/lindera-morphology/lindera/pull/188).
Upstream recently updated to that version (https://github.com/meilisearch/charabia/pull/105), so the next release of millisearch might fix this issue.

File: pkgs/servers/search/meilisearch/default.nix [hunk 1]
```diff
@@ -16,15 +16,18 @@ rustPlatform.buildRustPackage {
...
-  cargoPatches = [
-    # feature mini-dashboard tries to download a file from the internet
-    # feature analitycs should be opt-in
-    ./remove-default-feature.patch
-  ];
-  cargoSha256 = "sha256-dz+1IQZRSeMEagI2dnOtR3A8prg4UZ2Om0pd1BUhuhE=";
+  cargoSha256 = "sha256-+q5s/MndKFr8L6x40HYCHGhk5nm8KG8f+Lod9DeR+gQ=";
+  # this disables japanese search
+  # the japanese dependency is impure see https://github.com/lindera-morphology/lindera/blob/main/lindera-ipadic/build.rs
+  DOCS_RS=true;
```
Comment (SuperSandro2000) [inline pkgs/servers/search/meilisearch/default.nix:24]:
```suggestion
  DOCS_RS = true;
```

File: pkgs/servers/search/meilisearch/default.nix [hunk 2]
```diff
@@ -16,15 +16,18 @@ rustPlatform.buildRustPackage {
...
+  buildNoDefaultFeatures = true;
   buildInputs = lib.optionals stdenv.isDarwin [ Security DiskArbitration Foundation ];
+
+  # for some reason the tests don't even compile
```
Comment (SuperSandro2000) [inline pkgs/servers/search/meilisearch/default.nix:28]: Please add the most important part of the log to the comment.

**Comment by happysalada** [comment]
Thank you both for coming here to comment.
This still unfortunately has a rust build error that I couldn't get rid of. I'll be trying with every new version, but I'm not super hopeful.
Sandro, thank you for your comments, I'll be sure to address those when I get close to merging.

**Comment by SuperSandro2000** [comment]
Could it be that it requires a newer rust version?

**Comment by happysalada** [comment]
it's one of the dependencies that fails to compile.
When I posted something about it, it seemed it might have been related to the fact that we vendor dependencies. I tested compilation without vendoring and it was fine.
The dependency in question is quite old and hasn't received updates in a while, so it makes me think that this won't be solved anytime soon. However, I'll be re-trying with every new version that comes out to see if the problem is solved.
When I get something working, I'll take this MR out of draft.

**Comment by bbenno** [comment]
Today I tried to compile Meilisearch v0.28.1  (on x86_64-linux) and was surprised, it worked nearly out of the box and I fortunately did not come across the afore mentioned issues. I used the following package.

(Because I'm not very familiar with Darwin and there were some issues, I commented out the relevant things.)

```nix
{ stdenv
, lib
, rustPlatform
, fetchFromGitHub
#, Security
#, DiskArbitration
#, Foundation
}:

let version = "0.28.1";
in
rustPlatform.buildRustPackage {
  pname = "meilisearch";
  inherit version;
  src = fetchFromGitHub {
    owner = "meilisearch";
    repo = "MeiliSearch";
    rev = "v${version}";
    sha256 = "sha256-AjSzzOQ1K47tyc8Zn7kAU3B1UE9Tfvd3SvP7W13m6/o=";
  };
  cargoSha256 = "sha256-zRSAjOKKmL79eXrA6/ZNE3lo8MFdOqYJkcagcA51c6M=";
  buildNoDefaultFeatures = true;
  #buildInputs = lib.optionals stdenv.isDarwin [ Security DiskArbitration Foundation ];
  doCheck = false;

  meta = with lib; {
    broken = stdenv.isDarwin;
    description = "Powerful, fast, and an easy to use search engine ";
    homepage = "https://docs.meilisearch.com/";
    license = licenses.mit;
    maintainers = with maintainers; [ happysalada ];
    platforms = [ "x86_64-linux" "x86_64-darwin" ];
  };
}
```

**Comment by happysalada** [comment]
@bbenno thanks for coming by!
I've pushed a simple update of the version as a change.
Let's see what CI says.
In the previous CI run, it was failing on running the meilisearch nixos tests (for linux). There might have been an api change that causes the test to fail. If you have a moment to run the updated version and verify it works, it would be helpful.

**Comment by bbenno** [comment]
Hi @happysalada, thanks for the warm welcome.
I can verify that the updated version builds and works (at my x86_64-linux machine).

Although following warning occurs during rust compilation

```
[...]
   Compiling meilisearch-http v0.28.1 (/build/source/meilisearch-http)
warning: vergen: could not find repository from '/build/source/meilisearch-http'; class=Repository (6); code=NotFound (-3)
   Compiling lindera v0.13.5
[...]
```

**Comment by bbenno** [comment]
Following few (error) logs that might help to understand the package's values of `buildNoDefaultFeatures` and `doCheck`.

Each of the errors occurs, when
```
Compiling meilisearch-http v0.28.1 (/build/source/meilisearch-http)`
```

---

With ` doCheck = true` (and `buildNoDefaultFeatures = true`)

```
error: couldn't read /build/source/target/x86_64-unknown-linux-gnu/release/build/meilisearch-http-4f6b737478e94d90/out/generated.rs: No such file or directory (os error 2)
 --> meilisearch-http/tests/dashboard/mod.rs:8:9
  |
8 |         include!(concat!(env!("OUT_DIR"), "/generated.rs"));
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `include` (in Nightly builds, run with -Z macro-backtrace for more info)
  ```

---  

  When compiling with the [default features](https://github.com/meilisearch/meilisearch/blob/22aa349e31bc7662a065f4dc3229a93abd1f1f33/meilisearch-http/Cargo.toml#L90) not disabled (`buildNoDefaultFeatures = false`), it tries to download the meilisearch mini-dashboard repo during compilation process.

  ```
The following warnings were emitted during compilation:

warning: vergen: could not find repository from '/build/source/meilisearch-http'; class=Repository (6); code=NotFound (-3)

error: failed to run custom build command for `meilisearch-http v0.28.1 (/build/source/meilisearch-http)`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/meilisearch-http-b67736c40cd9403b/build-script-build` (exit status: 101)
  --- stdout
  cargo:warning=vergen: could not find repository from '/build/source/meilisearch-http'; class=Repository (6); code=NotFound (-3)

  --- stderr
  thread 'main' panicked at 'Could not load the mini-dashboard assets: error sending request for url (https://github.com/meilisearch/mini-dashboard/releases/download/v0.2.1/build.zip): error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution

  Caused by:
      0: error trying to connect: dns error: failed to lookup address information: Temporary failure in name resolution
      1: dns error: failed to lookup address information: Temporary failure in name resolution
      2: failed to lookup address information: Temporary failure in name resolution', meilisearch-http/build.rs:9:44
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  ```

**Comment by happysalada** [comment]
You're absolutely right, the no default features is to prevent the mini dashboard feature that downloads something from the internet. I'll add a comment to that affect to make it clearer to others.

The test phase failure appears to be the same. We could potentially try to disable default features for the test phase to try to run the tests.

If you look at the CI run (at the bottom of this PR, you will see that the nixos test fails with 
> Test "create index" failed with error: "index wasn't created"
)
I'm going to disable the nixos test for now so as to update this package.

If you are interested in either checking why the create index nixos test fail or trying out the check phase with disabling default features for it, you are welcome.

**Comment by bbenno** [comment]
Hi @happysalada, I think I misunderstood your question regarding the failing tests (thought you mean the `doCheck`, not the `passthru.tests` :see_no_evil: )

Yes, I think you are right with your assumption of an API Change breaking the test.

The error mentioned is raised by an [assertion, checking whether the creation of an index in a new setup meilisearch instance was successful](https://github.com/NixOS/nixpkgs/blob/0b4b412714e798056424b5e01e04a8bf462a9859/nixos/tests/meilisearch.nix#L41):

```
assert len(indexes) == 1, "index wasn't created"
```

However, the endpoint in use (`/indexes`) returns the following (with the index already created):
```
{'results': [{'uid': 'movies', 'createdAt': '2022-08-24T18:08:52.727732992Z', 'updatedAt': '2022-08-24T18:08:52.727732992Z', 'primaryKey': None}], 'offset': 0, 'limit': 20, 'total': 1}
```

Therefore, the assumption's `len(index)` returns `4`, causing the error.

---

The simplest fix would be to us the JSON's `total` attribute, representing the count of indexes.
I will open a PR right away.

**Pull request merged**